exporter/prometheus/prometheus.go

```diff
@@ -145,151 +115,164 @@ type collector struct {
+		desc := me.c.toDesc(metric)
+		me.descCh <- desc
 	}
+	return nil
 }
 
-func tagValues(t []tag.Tag, expectedKeys []tag.Key) []string {
-	var values []string
-	// Add empty string for all missing keys in the tags map.
-	idx := 0
-	for _, t := range t {
-		for t.Key != expectedKeys[idx] {
-			idx++
-			values = append(values, "")
-		}
-		values = append(values, t.Value)
-		idx++
-	}
-	for idx < len(expectedKeys) {
-		idx++
-		values = append(values, "")
+func toPromLabels(mls []string) (labels []string) {
+	for _, ml := range mls {
+		labels = append(labels, internal.Sanitize(ml))
 	}
-	return values
+	return labels
 }
 
-func viewName(namespace string, v *view.View) string {
+func metricName(namespace string, m *metricdata.Metric) string {
 	var name string
 	if namespace != "" {
 		name = namespace + "_"
 	}
-	return name + internal.Sanitize(v.Name)
+	return name + internal.Sanitize(m.Descriptor.Name)
+}
+
+func toPromMetric(
+	desc *prometheus.Desc,
+	metric *metricdata.Metric,
+	point metricdata.Point,
+	labelValues []string) (prometheus.Metric, error) {
+	switch metric.Descriptor.Type {
+	case metricdata.TypeCumulativeFloat64, metricdata.TypeCumulativeInt64:
+		pv, err := toPromValue(point)
+		if err != nil {
+			return nil, err
+		}
+		return prometheus.NewConstMetric(desc, prometheus.CounterValue, pv, labelValues...)
+
+	case metricdata.TypeGaugeFloat64, metricdata.TypeGaugeInt64:
+		pv, err := toPromValue(point)
+		if err != nil {
+			return nil, err
+		}
+		return prometheus.NewConstMetric(desc, prometheus.GaugeValue, pv, labelValues...)
+
+	case metricdata.TypeCumulativeDistribution:
+		switch v := point.Value.(type) {
+		case *metricdata.Distribution:
+			points := make(map[float64]uint64)
+			// Histograms are cumulative in Prometheus.
+			// Get cumulative bucket counts.
+			cumCount := uint64(0)
+			for i, b := range v.BucketOptions.Bounds {
+				cumCount += uint64(v.Buckets[i].Count)
+				points[b] = cumCount
+			}
+			return prometheus.NewConstHistogram(desc, uint64(v.Count), v.Sum, points, labelValues...)
+		default:
+			return nil, pointTypeError(point)
+		}
+	case metricdata.TypeSummary:
+		// TODO: [rghetia] add support for TypeSummary.
+		return nil, nil
+	default:
+		return nil, fmt.Errorf("aggregation %T is not yet supported", metric.Descriptor.Type)
+	}
 }
 
-func viewSignature(namespace string, v *view.View) string {
-	var buf bytes.Buffer
-	buf.WriteString(viewName(namespace, v))
-	for _, k := range v.TagKeys {
-		buf.WriteString("-" + k.Name())
+func toLabelValues(labelValues []metricdata.LabelValue) (values []string) {
+	for _, lv := range labelValues {
+		if lv.Present {
+			values = append(values, lv.Value)
+		} else {
+			values = append(values, "")
+		}
 	}
-	return buf.String()
+	return values
 }
 
-func (c *collector) cloneViewData() map[string]*view.Data {
-	c.mu.Lock()
-	defer c.mu.Unlock()
+func pointTypeError(point metricdata.Point) error {
+	return fmt.Errorf("point type %T is not yet supported", point)
```

Comment:
Nit: the error message could be a bit confusing, consider "point type %T doesn't match with metric type".

Reply:
done.